### DISCUSSION TITLE
Docker-image for Deployer

### DISCUSSION THREAD

**Discussion by alexgit2k** [deployphp/deployer#3893] [Features] [answered]
An official maintained Docker-image for Deployer would be helpful.

I found the following third party-images:
[~~pipelinecomponents/deployer~~](https://hub.docker.com/r/pipelinecomponents/deployer): version 7.3 (but seems only for Gitlab)
[~~minkbear/php-deployer~~](https://hub.docker.com/r/minkbear/php-deployer): version 7.1

Outdated:
[~~ttskch/deployer~~](https://hub.docker.com/r/ttskch/deployer): version 7.0
[~~mwienk/php-deployer~~](https://hub.docker.com/r/mwienk/php-deployer): version 6.8.0
[~~kblais/deployer~~](https://hub.docker.com/r/kblais/deployer): version 6.3.0
[~~herloct/deployer~~](https://hub.docker.com/r/herloct/deployer): version 5.0

**Comment by minkbear** [comment]
Here is my own build and used in gitlab CI. And just added 7.0.0-rc.8
https://hub.docker.com/r/minkbear/php-deployer/tags

**Comment by alexgit2k** [comment]
> Here is my own build and used in gitlab CI. And just added 7.0.0-rc.8 https://hub.docker.com/r/minkbear/php-deployer/tags

Great, I will add it to the list above.

Your tag [7.0.0-rc.8](https://hub.docker.com/layers/php-deployer/minkbear/php-deployer/7.0.0-rc.8/images/sha256-8ca4d1c93e0c6826f6e2cee281c0626076f66318dd2b88ae9b3608d319de7642?context=explore) uses PHP 7.4, it should use PHP 8.1, same for your tag [latest](https://hub.docker.com/layers/php-deployer/minkbear/php-deployer/latest/images/sha256-a9cb23eadf647a10f5f12f8ee98c1a6a82bdbccd34c6a81f2e43be83c5d1d259?context=explore).
Also a tag "7" for the latest Deployer-version 7 would be nice.

**Comment by minkbear** [comment]
> > Here is my own build and used in gitlab CI. And just added 7.0.0-rc.8 https://hub.docker.com/r/minkbear/php-deployer/tags
> 
> Great, I will add it to the list above.
> 
> Your tag [7.0.0-rc.8](https://hub.docker.com/layers/php-deployer/minkbear/php-deployer/7.0.0-rc.8/images/sha256-8ca4d1c93e0c6826f6e2cee281c0626076f66318dd2b88ae9b3608d319de7642?context=explore) uses PHP 7.4, it should use PHP 8.1, same for your tag [latest](https://hub.docker.com/layers/php-deployer/minkbear/php-deployer/latest/images/sha256-a9cb23eadf647a10f5f12f8ee98c1a6a82bdbccd34c6a81f2e43be83c5d1d259?context=explore). Also a tag "7" for the latest Deployer-version 7 would be nice.

Thanks for your recommendation so I reviewed and build the new one as described below

| Tag | Deployer Version | PHP Version |
| --- | --- | --- |
| `latest` | 6.8.0 | 8.1 |
| `6` | 6.8.0 | 8.1 |
| `7` | 7.0.0-rc.8 | 8.1 |

One problem is PHAR file is different url to download for example

URL of 7.0.0-beta.34 and below version is https://deployer.org/releases/v7.0.0-beta.34/deployer.phar 
but others are https://github.com/deployphp/deployer/releases/download/v7.0.0-beta.35/deployer.phar

@antonmedv is that possible to be the same URL.

**Comment by antonmedv** [comment]
Use first url, but follow redirects. curl -L

**Comment by minkbear** [comment]
> Use first url, but follow redirects. curl -L

Tested, Works fine. Thanks

**Comment by alexisLefebvre** [comment]
The doc uses the `deployphp/deployer:7` image: https://deployer.org/docs/7.x/ci-cd#gitlab-cicd It looks official.

**Comment by minkbear** [comment]
@alexislefebvre There is no image on docker hub (https://hub.docker.com/search?q=deployphp%2Fdeployer).

What am I missing?

**Comment by alexisLefebvre** [comment]
@minkbear You're right, it doesn't work, it can't pull this image:

```bash
docker pull deployphp/deployer:7
Error response from daemon: pull access denied for deployphp/deployer, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

---

Anyway, I suggest to use the [Docker registry of GitHub](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) which is easier to use than Docker Hub.

Here is an example: https://github.com/alexislefebvre/docker-images/blob/main/.github/workflows/push.yaml

GitHub build and publish the images through Actions, they are available through `ghcr.io/alexislefebvre/php-7.4-alpine-composer-gd`. It should be possible to add something like this to this repository in order to create Docker images on every release automatically.

**Comment by alexgit2k** [comment]
The docs are wrong, there is no deployer-image. See https://hub.docker.com/r/deployphp
Otherwise the maintainer would not have added the feature-flag here.

@alexislefebvre That's a nice solution for the deployer-maintainer or the maintainer of the other deployer-images.

**Comment by pknap** [comment]
Hi @minkbear,

since deployer 7.0.0 [now has been released](https://github.com/deployphp/deployer/releases/tag/v7.0.0), would it be possible that you updated your image to use 7.0.0 instead of 7.0.0-rc.8 on your [docker image](https://hub.docker.com/r/minkbear/php-deployer)?

**Comment by minkbear** [comment]
@pknap updated as your requested tag 7 is now 7.0.0 already.

| Tag | Deployer Version | PHP Version |
| --- | --- | --- |
| `latest` | 6.8.0 | 8.1 |
| `6` | 6.8.0 | 8.1 |
| `7` | 7.0.2 | 8.1 |
| `7.1` | 7.1.3 | 8.2 |

Thanks.

**Comment by minkbear** [comment]
@antonmedv got the problem that https://deployer.org/releases/v7.0.0/deployer.phar is not exists.

**Comment by qwerin** [comment]
> @antonmedv got the problem that https://deployer.org/releases/v7.0.0/deployer.phar is not exists.

https://github.com/deployphp/deployer/releases/download/v7.0.0/deployer.phar

**Comment by minkbear** [comment]
> https://deployer.org/releases/v7.0.0-beta.34/deployer.phar

Actually @antonmedv mention above (https://github.com/deployphp/deployer/issues/3118#issuecomment-1126878980) that use "https://deployer.org/releases/v7.0.0-beta.34/deployer.phar" URL then it will redirect to correct phar.

**Comment by dennismetz** [comment]
@minkbear 
I would like to use the image with gitlab, unfortunately I don't quite understand.
Could you create a short tutorial here - that would be great.

**Comment by alexgit2k** [comment]
> I would like to use the image with gitlab, unfortunately I don't quite understand. Could you create a short tutorial here - that would be great.

@dennismetz There is already a documentation for Gitlab: https://deployer.org/docs/7.x/ci-cd#gitlab-cicd
Haven't tried it, but it should work by just replacing `name: deployphp/deployer:7` with `name: minkbear/php-deployer:7`

**Comment by minkbear** [comment]
Updated

| Tag | Deployer Version | PHP Version |
| --- | --- | --- |
| `latest` | 6.8.0 | 8.1 |
| `6` | 6.8.0 | 8.1 |
| `7` | 7.1.3 | 8.2 |

**Comment by alexgit2k** [comment]
@minkbear Thanks, I updated the initial posting with the latest version.
As Deployer version 7 is out of beta, tag `latest` should point to 7.1.3. Tag `7` should point to the latest 7-version 7.1.3. So if you want to keep 7.0.2 it should be tag `7.0`.

**Comment by pknap** [comment]
Hi @minkbear, 

sorry to bother you again, but would it be possible to keep 7.0 version under separate tag, or have another one with 7.1 but keep the PHP on 8.1? Currently there is no combination with dep 7.x and php 8.1, so when the app does not yet support php 8.2, the image kind of gets backward incompatible.

Thank you

**Comment by minkbear** [comment]
@alexgit2k @pknap 
Updated. I will change latest later.

| Tag | Deployer Version | PHP Version |
| --- | --- | --- |
| `latest` | 6.8.0 | 8.1 |
| `6` | 6.8.0 | 8.1 |
| `7` | 7.1.3 | 8.2 |
| `7.0` | 7.0.2 | 8.1 |

**Comment by antonmedv** [comment]
Should we create automation to publish docker releases as well?

**Comment by alexgit2k** [comment]
> Should we create automation to publish docker releases as well?

That would save a lot of work, if docker images get created for new deployer-releases automatically.
But first at least one official deployer-docker-image would be great.

**Comment by alexisLefebvre** [comment]
It's possible to push images to the GitHub registry when a tag is released: https://github.com/binxio/push-to-registry-gh-action-workflow/blob/main/.github/workflows/push-to-registry.yaml (I didn't test it but it looks good)

**Comment by alexisLefebvre** [comment]
Please see #3487

**Comment by rvanlaak** [comment]
I can't seem to find the repo related to Dockerfile and builds for the images. @minkbear would you be willing to share these with us?

**Comment by alexgit2k** [comment]
> I can't seem to find the repo related to Dockerfile and builds for the images. @minkbear would you be willing to share these with us?

https://github.com/minkbear/php-deployer

**Comment by marvinhinz** [comment]
just ran into this, why is the deployphp/deployer image given as an example in the docs, when there is no such published package?

**Comment by antonmedv** [comment]
@marvinhinz I have plans to add it, but still didn’t found time for it.

**Comment by rvanlaak** [comment]
@antonmedv would there be anything we can help with? Already considered a base image?

**Comment by antonmedv** [comment]
I guess php8 as base image.

**Comment by Schrank** [comment]
I would assume if @antonmedv is fine with it, we should think about adding a github action to provide a docker image. github has it's own registry and technically we should be able to build an image with the new deployer version. 

cc @rvanlaak 

I'm neither into docker nor github actions, but if someone has any questions I'm happy to help

**Comment by Konafets** [comment]
@Schrank There is already a GH action -> https://github.com/deployphp/action

**Comment by Schrank** [comment]
@Konafets Thanks! Correct me, but this action runs deployer, doesn't it? I'm looking for a docker image which can be used in bitbucket, gitlab and other pipelines.

Or is it possible? That would be awesome <3

**Comment by Konafets** [comment]
@Schrank Seems a coincidence. Just posted that comment with a workaround as long there is no official docker image:

https://github.com/deployphp/deployer/issues/3738#issuecomment-2042746121

**Comment by uacode** [comment]
@Schrank What about this variant?

bitbucket-pipelines.yml

```image: php:8.2

pipelines:
  default:
    - parallel:
        - step:
            name: Deployment
            script:
              - apt-get update && apt-get install -y libssh2-1-dev git unzip
              - git clone https://github.com/php/pecl-networking-ssh2.git /usr/src/php/ext/ssh2 && docker-php-ext-install ssh2
              - docker-php-ext-install -j$(nproc) ssh2
              - curl -LO https://github.com/deployphp/deployer/releases/latest/download/deployer.phar
              - mv deployer.phar ./dep
              - chmod +x ./dep
              - ./dep --version
              - ./dep deploy test
```

**Comment by Schrank** [comment]
@Konafets No coincidence 😂 Your post was my start to push this topic :)

@uacode That does exactly what we want in bitbucket and this is what we do, nonetheless it would save some time to have a docker image and not install everyhing every time. Thanks for providing it!

**Comment by alexgit2k** [comment]
> @Konafets Thanks! Correct me, but this action runs deployer, doesn't it? I'm looking for a docker image which can be used in bitbucket, gitlab and other pipelines.

Just updated the initial ticket, the latest deployer-version 7.3.3 is only available at https://hub.docker.com/r/pipelinecomponents/deployer

It has only an example for Gitlab, so I do not know if it is running with Bitbucket or other pipelines.

**Comment by uacode** [comment]
@Schrank Tell me please, its possible to set deployment strategy in bitbucket-pipelines.yml not in deploy.php?

**Comment by uacode** [comment]
~22 sec No Docker https://prnt.sc/n1Edf82Dxdy7
vs
~2 sec - Docker https://prnt.sc/7bxIfE2m8LcR

```
image: uacode/deployer

pipelines:
  default:
    - parallel:
        - step:
            name: Deployment
            script:
#              - php --version
              - dep deploy test
```

**Comment by antonmedv** [comment, marked as answer]
### Deployer now has an official Docker image: https://hub.docker.com/r/deployphp/deployer

**Reply by Schrank** [reply]
May I ask wether this image is from "us" or from the "docker community"?

**Reply by antonmedv** [reply]
This docker image is from us! 🎉 
Now Deployer has automatic release of Docker image on release: [.github/workflows/docker.yml](https://github.com/deployphp/deployer/blob/master/.github/workflows/docker.yml)

**Reply by alexgit2k** [reply]
> ### Deployer now has an official Docker image: https://hub.docker.com/r/deployphp/deployer
Thanks antonmedev!

There is still a small issue for GitLab CI/CD: https://github.com/deployphp/deployer/discussions/4007
